### DISCUSSION TITLE
[RFC] Public `NextApiError` class with default status messages

### DISCUSSION THREAD

**Discussion by kripod** [vercel/next.js#17852] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

When an API route request calls a function which may not succeed, an additional value is required for tracking errors, e.g. a boolean below:

```js
import { getSession } from 'next-auth/client'

export default function handler(req, res) {
  async function authorize() {
    const session = await getSession({ req })
    if (!session) {
      res.status(401).end('Authentication required')
      return false
    }
    if (session.user.email !== 'admin@example.com') {
      res.status(403).end('Forbidden')
      return false
    }
    return true
  }

  if (req.method === 'DELETE') {
    const authorized = await authorize()
    if (authorized) {
      database.delete(/* Custom business logic... */)
    }
  }
}
```

## Describe the solution you'd like

As [used by some API util internals](https://github.com/vercel/next.js/blob/71d798ce889932311f2ef575c4c80b09ad57bb6f/packages/next/next-server/server/api-utils.ts#L128-L134), the `ApiError` class could short-circuit code execution and [send an HTTP response](https://github.com/vercel/next.js/blob/71d798ce889932311f2ef575c4c80b09ad57bb6f/packages/next/next-server/server/api-utils.ts#L99-L100) immediately as a result. The code above could be simplified as follows:

```js
import { NextApiError } from 'next/api'
import { getSession } from 'next-auth/client'

export default function handler(req, res) {
  async function authorize() {
    const session = await getSession({ req })
    if (!session) throw new NextApiError(401, 'Authentication required')
    if (session.user.email !== 'admin@example.com') throw new NextApiError(403)
  }

  if (req.method === 'DELETE') {
    await authorize() // Throws on error, blocking execution when necessary
    database.delete(/* Custom business logic... */)
  }
}
```

## Describe alternatives you've considered

I'm not sure if it's possible to break code execution throughout function calls without throwing errors.

## Additional context

I've created PR #17847 as a proof-of-concept, but exporting `NextApiError` from the main `next` module would raise the bundle size considerably. I propose using a separate entry point instead, e.g. `'next/api'`.

**Comment by bjohn465** [comment]
This would be very helpful. I'm working on a new Next.js project and wanted to create some functions that did some validation. If the validation fails, I'd like the function to abort the request with an error. I'd like to avoid requiring each API endpoint to check whether the function succeeded or failed.

I can get it to work right now, but it depends on accessing the internal `APIError` class:
```javascript
import { ApiError } from "next/dist/next-server/server/api-utils";

export function requireMinimumApiVersion(req, minimumVersion) {
  if (apiVersionNumber(req) < minimumVersion) {
    throw new ApiError(
      400,
      `API version v${minimumVersion} or higher required for this endpoint`
    );
  }
}
```

Relying on an internal error class is far from ideal, though, so an official way to achieve this would be great.

**Comment by kripod** [comment]
Also, throwing different errors for each HTTP status code might be something to consider for convenience, see: https://github.com/curveball/http-errors

@timneutkens I would love to hear your thoughts about integrating a package like @curveball/http-errors.
